src/Juvix/Compiler/Core/Language.hs

```diff
@@ -110,38 +142,51 @@ type Type = Node
 
 instance HasAtomicity Node where
   atomicity = \case
-    Var {} -> Atom
-    Ident {} -> Atom
-    Constant {} -> Atom
-    App {} -> Aggregate appFixity
-    BuiltinApp {..} | null _builtinArgs -> Atom
-    BuiltinApp {} -> Aggregate lambdaFixity
-    Constr {..} | null _constrArgs -> Atom
-    Constr {} -> Aggregate lambdaFixity
-    Lambda {} -> Aggregate lambdaFixity
-    Let {} -> Aggregate lambdaFixity
-    Case {} -> Aggregate lambdaFixity
-    Pi {} -> Aggregate lambdaFixity
-    Univ {} -> Atom
-    TypeConstr {} -> Aggregate appFixity
+    NVar {} -> Atom
+    NIdt {} -> Atom
+    NCst {} -> Atom
+    NApp {} -> Aggregate appFixity
+    NBlt BuiltinApp {..} | null _builtinAppArgs -> Atom
+    NBlt BuiltinApp {} -> Aggregate lambdaFixity
+    NCtr Constr {..} | null _constrArgs -> Atom
+    NCtr Constr {} -> Aggregate lambdaFixity
+    NLam Lambda {} -> Aggregate lambdaFixity
+    NLet Let {} -> Aggregate lambdaFixity
+    NCase Case {} -> Aggregate lambdaFixity
+    NPi Pi {} -> Aggregate lambdaFixity
+    NUniv Univ {} -> Atom
+    NTyp TypeConstr {} -> Aggregate appFixity
     Closure {} -> Aggregate lambdaFixity
 
 lambdaFixity :: Fixity
 lambdaFixity = Fixity (PrecNat 0) (Unary AssocPostfix)
 
 instance Eq Node where
   (==) :: Node -> Node -> Bool
-  Var _ idx1 == Var _ idx2 = idx1 == idx2
-  Ident _ sym1 == Ident _ sym2 = sym1 == sym2
-  Constant _ v1 == Constant _ v2 = v1 == v2
-  App _ l1 r1 == App _ l2 r2 = l1 == l2 && r1 == r2
-  BuiltinApp _ op1 args1 == BuiltinApp _ op2 args2 = op1 == op2 && args1 == args2
-  Constr _ tag1 args1 == Constr _ tag2 args2 = tag1 == tag2 && args1 == args2
-  Lambda _ b1 == Lambda _ b2 = b1 == b2
-  Let _ v1 b1 == Let _ v2 b2 = v1 == v2 && b1 == b2
-  Case _ v1 bs1 def1 == Case _ v2 bs2 def2 = v1 == v2 && bs1 == bs2 && def1 == def2
-  Pi _ ty1 b1 == Pi _ ty2 b2 = ty1 == ty2 && b1 == b2
-  Univ _ l1 == Univ _ l2 = l1 == l2
-  TypeConstr _ sym1 args1 == TypeConstr _ sym2 args2 = sym1 == sym2 && args1 == args2
+  NVar (Var _ idx1) == NVar (Var _ idx2) = idx1 == idx2
+  NIdt (Ident _ sym1) == NIdt (Ident _ sym2) = sym1 == sym2
+  NCst (Constant _ v1) == NCst (Constant _ v2) = v1 == v2
+  NApp (App _ l1 r1) == NApp (App _ l2 r2) = l1 == l2 && r1 == r2
+  NBlt (BuiltinApp _ op1 args1) == NBlt (BuiltinApp _ op2 args2) = op1 == op2 && args1 == args2
+  NCtr (Constr _ tag1 args1) == NCtr (Constr _ tag2 args2) = tag1 == tag2 && args1 == args2
+  NLam (Lambda _ b1) == NLam (Lambda _ b2) = b1 == b2
+  NLet (Let _ v1 b1) == NLet (Let _ v2 b2) = v1 == v2 && b1 == b2
+  NCase (Case _ v1 bs1 def1) == NCase (Case _ v2 bs2 def2) = v1 == v2 && bs1 == bs2 && def1 == def2
+  NPi (Pi _ ty1 b1) == NPi (Pi _ ty2 b2) = ty1 == ty2 && b1 == b2
+  NUniv (Univ _ l1) == NUniv (Univ _ l2) = l1 == l2
+  NTyp (TypeConstr _ sym1 args1) == NTyp (TypeConstr _ sym2 args2) = sym1 == sym2 && args1 == args2
   Closure _ env1 b1 == Closure _ env2 b2 = env1 == env2 && b1 == b2
   _ == _ = False
```

Comment:
I think this wildcard is dangerous. I suggest you we do something like this (taken from Inference.hs) to avoid it:
```
              case (normA, normB) of
                (ExpressionIden a, ExpressionIden b) -> goIden a b
                (ExpressionApplication a, ExpressionApplication b) -> goApplication a b
                (ExpressionFunction a, ExpressionFunction b) -> goFunction a b
                (ExpressionUniverse u, ExpressionUniverse u') -> check (u == u')
                (ExpressionLambda a, ExpressionLambda b) -> goLambda a b
                (ExpressionHole h, a) -> goHole h a
                (a, ExpressionHole h) -> goHole h a
                (ExpressionLambda {}, _) -> err
                (_, ExpressionLambda {}) -> err
                (ExpressionIden {}, _) -> err
                (_, ExpressionIden {}) -> err
                (ExpressionApplication {}, _) -> err
                (_, ExpressionApplication {}) -> err
                (ExpressionFunction {}, _) -> err
                (_, ExpressionFunction {}) -> err
                (ExpressionUniverse {}, _) -> err
                (_, ExpressionUniverse {}) -> err
                (ExpressionLiteral l, ExpressionLiteral l') -> check (l == l')
```

Reply:
Yes, let's do it.

Reply:
I used deriving stock (Eq) for Node itself